Engine/DLL/DLLManager.cpp

```diff
@@ -0,0 +1,188 @@
+#define CR_HOST CR_SAFE //must be here for HOT RELOADING library.
+#include "DLLManager.h"
+
+#include "Filesystem/Path.h"
+#include "Filesystem/PathAtlas.h"
+#include "Helper/Utils.h"
+#include "Main/Application.h"
+#include "Module/ModuleFileSystem.h"
+
+#include <thread>
+
+DLLManager::DLLManager()
+{
+
+#if GAME
+	//TODO USE THE NEW FILESYSTEM TO DO THIS
+	Utils::GetCurrentPath(working_directory);
+	working_directory += "/" + std::string(RESOURCE_SCRIPT_DLL_FILE);
+	CopyFile(RESOURCES_SCRIPT_DLL_PATH, working_directory.c_str(), false);
+	gameplay_dll = LoadLibrary(RESOURCE_SCRIPT_DLL_FILE);
+#else
+	CleanFolder();
+
+	dll_file = App->filesystem->GetPath(std::string("/") + RESOURCES_SCRIPT_DLL_PATH);
+	cr_plugin_open(hot_reloading_context, RESOURCES_SCRIPT_DLL_PATH);
+	cr_plugin_update(hot_reloading_context);
+	InitDLL();
+
+	init_timestamp_dll = dll_file->GetModificationTimestamp();
+
+	InitFolderTimestamps();
+#endif	
+
+}
+#if !GAME
+void DLLManager::InitFolderTimestamps()
+{
+	scripts_folder = App->filesystem->GetPath(std::string("/") + SCRIPT_PATH);
+	init_timestamp_script_folder = scripts_folder->GetModificationTimestamp();
+	scripts_timestamp_map.clear();
+	for (const auto& script : scripts_folder->children)
+	{
+		uint32_t timestamp = script->GetModificationTimestamp();
+		std::pair<Path*, uint32_t> script_timestamp (script, timestamp);
+		scripts_timestamp_map.insert(script_timestamp);
+	}
+}
+
+bool DLLManager::CheckFolderTimestamps()
+{
+	for (const auto& script : scripts_folder->children)
+	{
+		std::unordered_map<Path*, uint32_t>::const_iterator search = scripts_timestamp_map.find(script);
+		if (search == scripts_timestamp_map.end())
+		{
+			uint32_t timestamp = script->GetModificationTimestamp();
+			std::pair<Path*, uint32_t> script_timestamp(script, timestamp);
+			scripts_timestamp_map.insert(script_timestamp);
+			return true;
+		}
+		else 
+		{
+			uint32_t new_timestamp = search->first->GetModificationTimestamp();
+			if (search->second != new_timestamp)
+			{
+				scripts_timestamp_map[search->first] = new_timestamp;
+				return true;
+			}
+		}
+
+	}
+	return false;
+}
+
+bool DLLManager::DLLItsUpdated()
+{
+
+	last_timestamp_dll = dll_file->GetModificationTimestamp();
+	if (last_timestamp_dll != init_timestamp_dll)
+	{
+		init_timestamp_dll = last_timestamp_dll;
+		hot_reloading = true;
+		return true;
+	}
+
+	return false;
+}
+
+void DLLManager::CheckGameplayFolderStatus()
+{
+	last_timestamp_script_folder = scripts_folder->GetModificationTimestamp();
+	if (last_timestamp_script_folder != init_timestamp_script_folder || CheckFolderTimestamps())
+	{
+		std::thread(&DLLManager::CheckCompilation, this).detach();
+	}
+}
```

Comment:
Just a question, why are u calling CheckCompilation like &DLLManager::CheckCompilation on the thread and not CheckCompilation as is defined on the same class.

Reply:
If I'm not mistaken, std::thread in order to call a function that is member of a class, need to pass this function as it is there, creating a callback function. But to call a member function, we need a object and thats why also im passing "this".